meteor-bundle-main.js

```diff
@@ -73,29 +113,150 @@ function monkeypatchHttpAndHttps() {
           // hasn't given us a signed certificate yet, the response-json
           // won't exist.
           if ((e.code === 'EACCES') || (e.code == 'ENOENT')) {
-            console.log("Skipping unreadable HTTPS key information file:", metadataFilename);
+            console.log("Skipping unreadable HTTPS key information file while examining:",
+                        metadataFilename);
             continue;
           } else {
-            throw e;
+            // Sometimes the server gives us a 0-byte response,
+            // presumably due to a connection getting reset?.
+            console.error("Got exception reading JSON from file:", metadataFilename, "Dazed and confused, but trying to continue.", e);
+            continue;
           }
         }
 
-        // If this certificate isn't valid yet, keep looping, hoping
-        // to find one that is valid.
-        var notBefore = metadata['notBefore'];
-        if (notBefore && (notBefore < nowUnixTimestamp)) {
+        // If the cert is expired, definitely don't use it.
+        if (now > metadata.notAfter) {
+          console.log("Skipping", keyFilename, "because", metadata.notAfter, "is in the past.");
           continue;
         }
 
-        result['ca'] = metadata['ca'];
-        result['key'] = fs.readFileSync(keyFilename, 'utf-8');
-        result['cert'] = metadata['cert'];
-        return result;
+        validCertificates.push(metadata);
+      }
+
+      // Sort by notBefore, ascending.
+      validCertificates.sort(function(cert1, cert2) {
+        return (cert1.notBefore - cert2.notBefore);
+      });
+
+      if (! validCertificates) {
+        // TODO: Be more robust here, though it's not clear how.
+        console.error("Aieee, we found zero valid certificates. Allowing Sandstorm to crash.");
+        return;
       }
+
+      console.log("Using this HTTPS key:", validCertificates[0].keyFilename,
+                  "valid starting", new Date(validCertificates[0].notBefore),
+                  "until", new Date(validCertificates[0].notAfter));
+
+      // Store HTTPS configuration data from the oldest certificate.
+      var result = {
+        ca: validCertificates[0].ca,
+        key: fs.readFileSync(validCertificates[0].keyFilename, 'utf-8'),
+        cert: validCertificates[0].cert
+      };
+
+      // Calculate re-key time.
+      //
+      // - If the cert we want to switch to is not valid yet, then we
+      //   re-key at its notBefore time.
```

Comment:
Yeah. I was trying to think through the best algorithm here, and wanted to deal with that after submitting this basic implementation.

So thinking aloud:

* Scenario:
    * Sandstorm is already running, and
    * cert 1 expires tomorrow 12:00 PM
    * cert 2 starts today 12:00 PM
    * time right now is 12:00 PM

* Clock that is slow - thinks it's 11:55 AM
    * We keep using cert 1 until "12:00 PM" which is actually 12:05 PM
    * We switch to cert 2 at "12:00 PM" actually 12:05 PM
    * Result: This is OK (because cert 1 didn't expire yet).

* Clock that is fast - thinks it's 12:05 PM
    * We stopped using cert 1 at "12:00 PM" which was actually "11:55 AM"
    * We switched to cert 2 at "12:00 PM" actually 11:55 AM
    * Result: For five minutes, web clients see an expired certificate.
    * Proposal to fix this: Do the switch-over at the _average_ of the cert 1 expiry & cert 2 start date.
    * Alternate proposal: Do the switch-over some constant factor into cert 2 (e.g. 10 minutes) so long as cert 1 is still valid for that time.